File: circly-app/src/types/poll.ts
// Removed UserResponse import as it's not used

// Poll Option types
export interface PollOption {
  id: string; // UUID
  poll_id: string;
  member_id: number;
  member_nickname: string;
  display_order: number;
  vote_count: number;
  created_at: string;
}

export interface PollOptionCreate {
  text: string;
  user_id?: number;
  order_index?: number;
}

// Poll types
export interface Poll {
  id: number;
  title: string;
  description?: string;
  question_template: string;
  circle_id: number;
  creator_id: number;
  expires_at?: string;
  is_active: boolean;
  is_anonymous: boolean;
  created_at: string;
  updated_at?: string;
}

export interface PollResponse {
  id: string; // UUID
  question_text: string; // 백엔드의 question_text
  circle_id: number;
  circle_name?: string; // Circle 이름 (선택적)
  creator_id: number;
  template_id?: string;
  deadline: string; // 백엔드의 deadline
  is_active: boolean;
  is_anonymous: boolean;
  is_closed: boolean;
  max_votes_per_user: number;
  total_votes: number;
  total_participants: number;
  created_at: string;
  options: PollOption[];
  user_voted: boolean;
  user_vote_option_id?: string;
}

export interface PollCreate {
  title: string;
  description?: string;
  question_template: string;
  circle_id: number;
  expires_at?: string;
  is_anonymous?: boolean;
  options: PollOptionCreate[];
}

export interface PollUpdate {
  title?: string;
  description?: string;
  expires_at?: string;
  is_active?: boolean;
}

// Vote types
export interface Vote {
  id: number;
  poll_id: number;
  option_id: number;
  user_id: number;
  voted_at: string;
}

export interface VoteCreate {
  option_id: string; // UUID
}

export interface VoteResult {
  option_id: string; // UUID
  vote_count: number;
  percentage: number;
}

// Poll participation types
export interface PollParticipation {
  poll_id: string; // UUID
  has_voted: boolean;
  selected_option_id?: string;
  voted_at?: string;
}

// Poll state
export interface PollState {
  polls: PollResponse[];
  currentPoll: PollResponse | null;
  loading: boolean;
  error: string | null;
}

// Poll template types
export interface PollTemplate {
  id: string;
  title: string;
  question: string;
  category: string;
  description?: string;
  defaultOptions?: string[];
}

File: backend/app/schemas/poll.py
from pydantic import BaseModel, Field, validator
from typing import Optional, List
from datetime import datetime


class PollOptionBase(BaseModel):
    member_id: int
    member_nickname: str
    display_order: int


class PollOptionCreate(PollOptionBase):
    pass


class PollOptionResponse(PollOptionBase):
    id: str
    poll_id: str
    vote_count: int
    created_at: datetime

    class Config:
        from_attributes = True


class PollBase(BaseModel):
    question_text: str = Field(..., min_length=5, max_length=500)
    deadline: datetime


class PollCreate(PollBase):
    template_id: str
    circle_id: int
    
    @validator('deadline')
    def deadline_must_be_future(cls, v):
        if v <= datetime.now():
            raise ValueError('마감 시간은 현재 시간보다 미래여야 합니다')
        return v


class PollUpdate(BaseModel):
    question_text: Optional[str] = None
    deadline: Optional[datetime] = None
    is_active: Optional[bool] = None


class PollResponse(PollBase):
    id: str
    circle_id: int
    creator_id: int
    template_id: Optional[str]
    is_anonymous: bool
    max_votes_per_user: int
    is_active: bool
    is_closed: bool
    total_votes: int
    total_participants: int
    created_at: datetime
    options: List[PollOptionResponse] = []

    class Config:
        from_attributes = True


class PollListResponse(BaseModel):
    polls: List[PollResponse]
    total: int
    limit: int
    offset: int


class VoteCreate(BaseModel):
    option_id: str


class VoteResponse(BaseModel):
    id: str
    poll_id: str
    option_id: str
    created_at: datetime

    class Config:
        from_attributes = True


class PollResultOption(BaseModel):
    option_id: str
    member_nickname: str
    vote_count: int
    percentage: float
    rank: int


class PollResultResponse(BaseModel):
    poll_id: str
    question_text: str
    total_votes: int
    total_participants: int
    is_closed: bool
    deadline: datetime
    results: List[PollResultOption]
    winner: Optional[PollResultOption] = None


class PollWithUserStatus(PollResponse):
    user_voted: bool
    user_vote_option_id: Optional[str] = None